File: contest6/pE.cpp
#include <bits/stdc++.h>
#define ll long long int
using namespace std;

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    // freopen("input.txt", "r", stdin);
    /// bibliografia: guide to competitive programming
    /*
        revisar:
        getline() ///hay distintos parametros aqui.

    */
    map<ll, ll> twos, rep;
    twos[0] = 1ll;
    for (ll i = 1; i < 30; i++)
        twos[i] = 2ll * twos[i - 1ll];
    ll m, op, val;
    cin >> m;
    for (ll i = 0; i < m; i++) {
        cin >> op >> val;
        if (op == 1) {
            rep[twos[val]]++;
        } else {
            for (ll j = 29; j >= 0; j--) {
                if (val >= twos[j] && val) {
                    if (rep[twos[j]])
                        val -= min(val / twos[j], rep[twos[j]]) * twos[j];
                }
            }
            if (!val)
                cout << "YES" << "\n";
            else
                cout << "NO" << "\n";
        }
    }
    return 0;
}

File: contest6/pE.py
twos = {}
rep = {}
twos[0] = 1
for i in range(1, 30):
    twos[i] = 2 * twos[i - 1]
m = int(input())
for i in range(m):
    op, val = map(int, input().split())
    if op == 1:
        if twos[val] in rep:
            rep[twos[val]] += 1
        else:
            rep[twos[val]] = 1
    else:
        for j in range(29, -1, -1):
            if val >= twos[j] and val:
                if twos[j] in rep:
                    val -= min(val // twos[j], rep[twos[j]]) * twos[j]
        if not val:
            print("YES")
        else:
            print("NO")
